File: core/src/gregdustry/world/blocks/MultiFactory.java
package gregdustry.world.blocks;

import arc.struct.Seq;
import arc.util.Time;
import arc.util.io.Reads;
import arc.util.io.Writes;
import mindustry.Vars;
import mindustry.ctype.UnlockableContent;
import mindustry.gen.Building;
import mindustry.type.Item;
import mindustry.type.Liquid;
import mindustry.world.Block;

public class MultiFactory extends Block {
    // This public field matches your Hjson key name perfectly
    public Seq<RecipeData> productionList = new Seq<>();
    public Seq<ParsedRecipe> parsedRecipes = new Seq<>();

    public MultiFactory(String name) {
        super(name);
        update = true;
        configurable = true;
        hasItems = true;
        hasLiquids = true;
    }

    // Temporary layout structures for the Hjson reader
    public static class RecipeData {
        public String produces;
        public int speed = 60;
        public int amount = 1;
        public Seq<InputData> using = new Seq<>();
    }

    public static class InputData {
        public String name;
        public int amount = 1;
        public boolean allowFlammableAlternatives = false;
    }

    // Cleaned-up recipe structures for runtime logic loops
    public static class ParsedRecipe {
        public int id;
        public UnlockableContent output;
        public int amount;
        public boolean isLiquid;
        public float craftTime;
        public Seq<InputData> inputs;
    }

    @Override
    public void init() {
        super.init();

        // Convert the raw Hjson string data into real game Content objects
        for (int i = 0; i < productionList.size; i++) {
            RecipeData raw = productionList.get(i);
            ParsedRecipe recipe = new ParsedRecipe();
            recipe.id = i;

            // Look up items first, then fall back to liquids
            recipe.output = Vars.content.item(raw.produces);
            if (recipe.output == null) {
                recipe.output = Vars.content.liquid(raw.produces);
                recipe.isLiquid = true;
            }

            recipe.amount = raw.amount;
            recipe.craftTime = raw.speed;
            recipe.inputs = raw.using;

            if (recipe.output != null) {
                parsedRecipes.add(recipe);
            }
        }
    }



    public class MultiFactoryBuilding extends Building {
        public int recipeId = -1;
        public float progress = 0;

        @Override
        public void updateTile() {
            if (recipeId == -1 || recipeId >= parsedRecipes.size) return;
            ParsedRecipe recipe = parsedRecipes.get(recipeId);

            if (hasInputs(recipe)) {
                float boostMultiplier = 1.0f;

                for (InputData req : recipe.inputs) {
                    if (req.allowFlammableAlternatives) {
                        Item usedItem = getBestFlammableItem(req.name);
                        if (usedItem != null) {
                            boostMultiplier = Math.max(boostMultiplier, usedItem.flammability);
                        }
                    }
                }

                // Tick progress calculation
                progress += (Time.delta * boostMultiplier) / recipe.craftTime;

                if (progress >= 1f) {
                    consumeInputs(recipe);
                    produceOutput(recipe);
                    progress = 0f;
                }
            }
        }

        public boolean hasInputs(ParsedRecipe recipe) {
            for (InputData req : recipe.inputs) {
                Item baseItem = Vars.content.item(req.name);
                if (baseItem != null && items.get(baseItem) >= req.amount) continue;

                if (req.allowFlammableAlternatives) {
                    boolean substituteFound = false;
                    for (int i = 0; i < Vars.content.items().size; i++) {
                        Item item = Vars.content.items().get(i);
                        if (item.flammability > 0 && items.get(item) >= req.amount) {
                            substituteFound = true;
                            break;
                        }
                    }
                    if (substituteFound) continue;
                }
                return false;
            }
            return true;
        }

        public void consumeInputs(ParsedRecipe recipe) {
            for (InputData req : recipe.inputs) {
                Item item = Vars.content.item(req.name);
                if (item != null && items.get(item) >= req.amount) {
                    items.remove(item, req.amount);
                } else if (req.allowFlammableAlternatives) {
                    Item sub = getBestFlammableItem(req.name);
                    if (sub != null) items.remove(sub, req.amount);
                }
            }
        }

        public void produceOutput(ParsedRecipe recipe) {
            if (!recipe.isLiquid) {
                this.handleStack((Item) recipe.output, recipe.amount, this);
            } else {
                // Correct v158.1 method for adding liquids directly to the building
                this.handleLiquid(this, (Liquid) recipe.output, recipe.amount);
            }
        }

        public Item getBestFlammableItem(String baseName) {
            Item base = Vars.content.item(baseName);
            if (base != null && items.get(base) > 0) return base;

            Item best = null;
            for (int i = 0; i < Vars.content.items().size; i++) {
                Item item = Vars.content.items().get(i);
                if (item.flammability > 0 && items.get(item) > 0) {
                    if (best == null || item.flammability > best.flammability) {
                        best = item;
                    }
                }
            }
            return best;
        }

        // Network and Save Data Packet Serialization
        @Override
        public void write(Writes write) {
            super.write(write);
            write.i(recipeId);
            write.f(progress);
        }

        @Override
        public void read(Reads read, byte revision) {
            super.read(read, revision);
            recipeId = read.i();
            progress = read.f();
        }

        @Override
        public Object config() {
            return recipeId;
        }
    }
}

File: assets/scripts/multi-factory.js
Events.on(ClientLoadEvent, () => {
    // Get all blocks registered in the game
    let allBlocks = Vars.content.blocks();

    for (let i = 0; i < allBlocks.size; i++) {
        let block = allBlocks.get(i);

        // Filter: Only target blocks belonging to your mod that contain a productionList
        if (block.name.startsWith("gregdustry-") && block.customData) {
            initializeMultiFactory(block);
            Log.info(`[Gregdustry] ${block.name} Loaded as MultiFactory`)
        }
    }
    Log.info("[Gregdustry] Dynamically initialized all Multi-Factory components.");
});

function initializeMultiFactory(block) {
    // 1. Force structural parameters on the target block configuration
    block.configurable = true;
    block.hasItems = true;
    block.hasLiquids = true;
    block.hasPower = true;
    block.update = true;

    // 2. Parse the recipe array mapped from the Hjson structure
    block.parsedRecipes = [];
    let list = block.productionList;
    
    for (let i = 0; i < list.length; i++) {
        let raw = list[i];
        block.parsedRecipes.push({
            id: i,
            output: Vars.content.item(raw.produces) || Vars.content.liquid(raw.produces),
            outputAmount: raw.amount || 1,
            isLiquidOutput: Vars.content.liquid(raw.produces) != null,
            craftTime: raw.speed || 60,
            inputs: raw.using || []
        });
    }

    // 3. Attach the selection UI button matrix
    block.buildConfiguration = function(tile, table) {
        block.parsedRecipes.forEach(recipe => {
            if (!recipe.output) return;
            table.button(new TextureRegionDrawable(recipe.output.uiIcon), Styles.cleari, () => {
                tile.configure(recipe.id);
            }).size(50).pad(4);
        });
    };

    // 4. Overwrite the entity building instance runtime loops
    block.buildType = () => extend(Building, {
        _recipeId: -1,
        _progress: 0,

        configure(value) {
            this._recipeId = value;
            this._progress = 0;
        },

        config() {
            return java.lang.Integer.valueOf(this._recipeId);
        },

        updateTile() {
            if (this._recipeId === -1) return;
            let recipe = block.parsedRecipes[this._recipeId];
            
            if (this.hasInputs(recipe)) {
                let boostMultiplier = 1.0;
                
                recipe.inputs.forEach(req => {
                    if (req.allowFlammableAlternatives) {
                        let usedItem = this.getBestFlammableItem(req.name);
                        if (usedItem) {
                            boostMultiplier = Math.max(boostMultiplier, usedItem.flammability);
                        }
                    }
                });

                // Delta time frame step processing
                this._progress += (Time.delta * boostMultiplier) / recipe.craftTime;
                
                if (this._progress >= 1) {
                    this.consumeInputs(recipe);
                    this.produceOutput(recipe);
                    this._progress = 0;
                }
            }
        },

        getBestFlammableItem(baseName) {
            let base = Vars.content.item(baseName);
            if (this.items.get(base) > 0) return base;
            
            let best = null;
            for (let i = 0; i < Vars.content.items().size; i++) {
                let item = Vars.content.items().get(i);
                if (item.flammability > 0 && this.items.get(item) > 0) {
                    if (!best || item.flammability > best.flammability) best = item;
                }
            }
            return best;
        },

        hasInputs(recipe) {
            for (let i = 0; i < recipe.inputs.length; i++) {
                let req = recipe.inputs[i];
                let baseItem = Vars.content.item(req.name);
                
                if (baseItem && this.items.get(baseItem) >= req.amount) continue;

                if (req.allowFlammableAlternatives) {
                    let substituteFound = false;
                    for (let j = 0; j < Vars.content.items().size; j++) {
                        let item = Vars.content.items().get(j);
                        if (item.flammability > 0 && this.items.get(item) >= req.amount) {
                            substituteFound = true;
                            break;
                        }
                    }
                    if (substituteFound) continue;
                }
                return false;
            }
            return true;
        },

        consumeInputs(recipe) {
            recipe.inputs.forEach(req => {
                let item = Vars.content.item(req.name);
                if (item) {
                    this.items.remove(item, req.amount);
                } else if (req.allowFlammableAlternatives) {
                    let sub = this.getBestFlammableItem(req.name);
                    if (sub) this.items.remove(sub, req.amount);
                }
            });
        },

        produceOutput(recipe) {
            if (!recipe.isLiquidOutput) {
                this.handleStack(recipe.output, recipe.outputAmount, this);
            } else {
                this.liquids.handleStack(recipe.output, recipe.outputAmount, this);
            }
        }
    });

    Log.info("[Gregdustry] Configured multi-factory module for block: " + block.name);
}